crates/torii/core/src/engine.rs

```diff
@@ -317,9 +288,110 @@ impl<P: Provider + Sync> Engine<P> {
 
         debug!("Transactions: {}", &transactions.len());
         debug!("Blocks: {}", &blocks.len());
+
+        Ok(FetchRangeResult { transactions, blocks, latest_block_number: to })
+    }
+
+    async fn fetch_pending(
+        &self,
+        block_number: u64,
+        pending_block_tx: Option<Felt>,
+    ) -> Result<Option<FetchPendingResult>> {
+        let block = if let MaybePendingBlockWithTxs::PendingBlock(pending) =
+            self.provider.get_block_with_txs(BlockId::Tag(BlockTag::Pending)).await?
+        {
+            pending
+        } else {
+            // TODO: change this to unreachable once katana is updated to return PendingBlockWithTxs
+            // when BlockTag is Pending unreachable!("We requested pending block, so it
+            // must be pending");
+            return Ok(None);
+        };
+
+        Ok(Some(FetchPendingResult {
+            pending_block: Box::new(block),
+            block_number,
+            pending_block_tx,
+        }))
+    }
+
+    pub async fn process(&mut self, fetch_result: FetchDataResult) -> Result<()> {
+        match fetch_result {
+            FetchDataResult::Range(data) => {
+                self.process_range(data).await?;
+            }
+            FetchDataResult::Pending(data) => {
+                self.process_pending(data).await?;
+            }
+            FetchDataResult::None => {}
+        }
+
+        Ok(())
+    }
+
+    pub async fn process_pending(&mut self, data: FetchPendingResult) -> Result<()> {
+        // Skip transactions that have been processed already
+        // Our cursor is the last processed transaction
+        let mut pending_block_tx_cursor = data.pending_block_tx;
+        let mut pending_block_tx = data.pending_block_tx;
```

Comment:
We don't need to check the block too? If the pending goes fast, we might have a race by fetching the pending, processing, but the pending block has changed, so the tx hash will never be found?

Reply:
Should not happen with the current network speed, but wondering if it would be safer to ensure we don't mix data up? Mostly when we fetch just before the pending is mined for instance.